tratto/src/main/java/star/tratto/util/StringUtils.java

```diff
@@ -85,4 +128,143 @@ public static String fullyQualifiedClassName(String packageName, String classNam
+     */
+    private static Map<String, Integer> getHistogram(List<String> strings) {
+        Map<String, Integer> histogram = new HashMap<>();
+        for (String string : strings) {
+            int currentCount = histogram.getOrDefault(string, 0);
+            histogram.put(string, currentCount + 1);
+        }
+        return histogram;
+    }
+
+    /**
+     * Converts a histogram of word frequencies to a vector.
+     *
+     * @param frequencies a map of word frequencies
+     * @param words the set of all words to be considered in the vector
+     * @return a vector representation of the word frequencies. Each entry
+     * corresponds to a different word, where the value of the entry
+     * corresponds to the word frequency. If a word does not appear in the
+     * histogram, then it is assigned a value of 0. The length of the vector
+     * is equal to the number of words, {@code words.size()}.
+     */
+    private static RealVector wordFrequencyToVector(Map<String, Integer> frequencies, SortedSet<String> words) {
+        double[] vector = new double[words.size()];
+        int i = 0;
+        for (String word : words) {
+            vector[i] = frequencies.getOrDefault(word, 0);
+            i++;
+        }
+        return new ArrayRealVector(vector);
+    }
+
+    /**
+     * Returns the words in both sets.
+     *
+     * @param set1 a set of words
+     * @param set2 a set of words
+     * @return the words in both sets 
+     */
+    private static SortedSet<String> setIntersection(Set<String> set1, Set<String> set2) {
+        TreeSet<String> intersectionKeys = new TreeSet<>(set1);
+        intersectionKeys.retainAll(set2);
+        return intersectionKeys;
+    }
+
+    /**
+     * Computes the cosine similarity of two vectors. Imitates behavior of
+     * {@link RealVector#cosine} without throwing exceptions. If the norm of
+     * either vector is 0.0, then the method returns 0.0.
+     *
+     * @param vector1 a vector
+     * @param vector2 a vector
+     * @return the cosine similarity of the two vectors
+     */
+    private static double cosineSimilarity(RealVector vector1, RealVector vector2) {
+        double denominator = vector1.getNorm() * vector2.getNorm();
+        if (denominator == 0.0) {
+            return 0.0;
+        }
+        return vector1.dotProduct(vector2) / denominator;
+    }
+
+    /**
+     * Computes the cosine similarity of two lists of Strings.
+     *
+     * @param strings1 list of strings
+     * @param strings2 list of strings
+     * @return the cosine similarity (double between 0.0 and 1.0)
+     */
+    private static double cosineSimilarity(List<String> strings1, List<String> strings2) {
+        Map<String, Integer> wordsFreq1 = getHistogram(strings1);
+        Map<String, Integer> wordsFreq2 = getHistogram(strings2);
+        SortedSet<String> intersectionKeys = setIntersection(wordsFreq1.keySet(), wordsFreq2.keySet());
```

Comment:
References on the Web state to use the union of all words in both sentences.  Why is it correct to use the intersection here?